draft-nygren-httpbis-httpssvc.md

```diff
@@ -0,0 +1,791 @@
+network from implanting a forged DNS record that allows them to
+track users after they leave that network.
+
+## Multiple records and preference ordering
+
+Server operators MAY publish multiple HTTPSSVC records as an RRSET.  When
+publishing an RRSET with multiple HTTPSSVC
+records, the server operator MUST set the overall TTL to the minimum
+of the "max age" values (following Section 5.2 of {{!RFC2181}}).
+
+Each RR MUST contain exactly one alt-value, as described
+in Section 3 of {{!AltSvc}}.
+
+As RRs within an RRSET are explictly unordered collections, a new
+"pri" alt-svc parameter is introduced to indicate priority.  The value
+of the "pri" parameter is a non-negative integer between 0 and 65535.
+Alt-Svc alt-values with a smaller pri value should be given preference
+over alt-values with a larger pri value.
+
+TBD: what should default pri values be for Alt-Svc records
+received via DNS and HTTP when pri is unspecified?
+
+When receiving an RRSET containing multiple HTTPSSVC records with the
+same "pri" value, clients SHOULD apply a random shuffle within a
+priority level to the records before using them, to ensure randomized
+load-balancing.
+
+## Constructing Alt-Svc equivalent headers
+
+For a client to construct the equivalent of an Alt-Svc HTTP response header:
+
+1. For each RR, the SvcDomainName MUST be inserted as the uri-host.
+2. The RRs SHOULD be ordered by "pri" values, with shuffling
+   for equal "pri" values.  Clients MAY choose to further
+   prioritize records where address records are immediately
+   available for the record's SvcDomainName.
+3. The client SHOULD concatenate the thus-transformed-and-ordered SvcFieldValues
+   in the RRSET, separated by commas.  (This is semantically equivalent to
+   receiving multiple Alt-Svc HTTP response headers, according to Section 3.2.2
+   of {{?HTTP=RFC7230}}).
+
+## Granularity and lifetime control
+
+Sending Alt-Svc over HTTP allows the server to tailor the Alt-Svc
+Field Value specifically to the client.  When using an HTTPSSVC DNS
+record, groups of clients will necessarily receive the same Alt-Svc
+Field Value.  Therefore, this standard is not suitable for servers that
+require single-client granularity in Alt-Svc.  Server operators that
+want to serve different Alt-Svc Field Values to different geographic
+or network regions SHOULD configure their authoritative DNS server to
+respect the EDNS0 Client Subnet extension {{?RFC7871}}.
+
+Some DNS caching systems incorrectly extend the lifetime of DNS
+records beyond the stated TTL.  Server operators MUST NOT rely on
+HTTPSSVC records expiring on time, and MAY shorten the TTL to compensate.
+
+Clients SHOULD (MAY?) prefer alternative service records
+received over HTTPS to those received via DNS.
+
+
+# Client behaviors
+
+TODO: Write more details on the recommended resolution algorithm.
+This section does NOT yet cover the SvcRecordType=0 case well yet.
+
+
+## HTTP Strict Transport Security
```

Comment:
I’m leery of making this HSTS-equivalent, especially since it would be trivial to permit 2.2 to indicate that the https scheme can be served over HTTPS using existing Alt-Svc mechanisms.  Just point to RFC 8164 and keep going.

Reply:
I think HSTS support here is important.  It's a rare chance to fix the "bare domain" bootstrap problem, and would win a lot of support for HTTPSSVC from parties who would otherwise not be very interested in this stuff.

I don't view opportunistic security as sufficient, and I think it's considerably more complicated than "HTTPSSVC requires HTTPS".

Reply:
I tend to agree wtih Ben here that we should at least try and propose this.